### DISCUSSION TITLE
Greek combining characters

### DISCUSSION THREAD

**Discussion by ichnilatis-gr** [koreader/koreader#10082] [Q&A]
I would like to ask if anyone knows how or where I can find the Greek combining characters for Greek accents and breathings (such as ◌͂ and ◌̈́ and such as the characters that are in the pop-up menu of the English keyboard at number 6)?  

Thank you!

**Comment by poire-z** [comment]
> how or where I can find the Greek combining characters for Greek accents 

What do you mean by "how and where" ?

You want to find a complete reference of them on the web ?  Like https://en.wikipedia.org/wiki/Greek_diacritics or https://en.wikipedia.org/wiki/Combining_Diacritical_Marks ?

Or you want to find how to input them in KOreader ? If this, I don't know, they probably don't _have to_ be anywhere in the English keyboard (and derived ones). But I hope the author of the Greek keyboard (which is yourself :) has added them somewhere, in some panels or some popups :)

Or something else ?

**Comment by mergen3107** [comment]
You can apply any accent to any symbols:
1) Input symbol
2) Long-press the 6, tap the desired accent. 
3) If needed, tap another accent. 
4) The accents will start appearing above the character as expected.

**Comment by ichnilatis-gr** [comment]
Sorry, I wasn't clear! I want to make a pop-up menu like the one for the number 6 on the English keyboard with Greek combining characters for Greek accents and breathings. But I don't know where to copy them from. Any idea?

**Reply by poire-z** [reply]
Well, you'll find the Unicode code for these in the above Wikipedia urls, or links from them.
ie. <html><body>


U+0342 | ͂ | Combining Greek Perispomeni
-- | -- | --



</body>
</html>

Often, you'll see them presented with a dotted circle ◌͂ , which makes it harder to copy and remove the circle.
So, you may need to input the character from its Unicode code , ie 0342 - and the way to do this may depend on your OS (Windows, MacOS...). I don't even know how I would do it on my Windows :)

I also dunno if we should include the dotted circle in the keypopup, but @mergen3107 who did the keyboard revamp may know better about that.

**Reply by mergen3107** [reply]
I took all the circle-diacritics from the Unicode tables online, copying with “Copy” button. Look for “modifier” - try them out first on PC. If they modify this properly, then you can add them to KOReader’s keyboard files. 

Start here, fetch all your working diacritics you want and I can help you better later.

**Reply by ichnilatis-gr** [reply]
What website did you get them from?

**Reply by mergen3107** [reply]
Don’t remember, that was a few years ago :D 
It doesn’t matter, Unicode is Unicode, just try it

**Reply by ichnilatis-gr** [reply]
Unfortunately, I can't find all the combinations I want. I mean combinations with more than one diacritical mark. Is it possible that they are not in the tables, but only with specific letters?

**Reply by poire-z** [reply]
> combinations with more than one diacritical mark

The point with diacritical unicode codepoints is that when you need "more than one diacritical", you just combine them :)
Base_letter + diacritical1 + diacritical2

![image](https://user-images.githubusercontent.com/24273478/235357082-02a9e954-5703-4379-b41b-5386a69e627d.png)
(From https://fr.wikipedia.org/wiki/Normalisation_Unicode).

It may happen that sometimes, there is an existing unicode codepoint that carries some diacritical with its letter, because it is common:
![image](https://user-images.githubusercontent.com/24273478/235357276-47af148d-5c81-405a-90bb-637a24127ede.png)
but if you don't find it, you just create it with multiple codepoints. (But if you find it, I think it is preferable to use the single one.)

Your keyboard key should then output all the codepoint involved (it doesn't have to be single).

**Reply by ichnilatis-gr** [reply]
Oh! I didn't know that and didn't think to try it. That's perfect; no problem then! Well, I'll try it. Thank you very much.

**Reply by poire-z** [reply]
Btw, as you could enter them in any order, ie for the é. sample above:
```
e + ' + .    or
e + . + '
```
I think it's best to try to order them in the NFC form (so what we output is properly canonical ordered canonically Unicode :) @Frenzie : do you agree?

This may help you if you can input your final Greek heavily diacriticized characters on your computer:
https://minaret.info/test/normalize.msp
![image](https://user-images.githubusercontent.com/24273478/235460985-d3c4f33f-27fc-47b6-9a57-116fcb0aa685.png)

**Reply by ichnilatis-gr** [reply]
New difficulties... I'll try to understand this... :-)

**Reply by Frenzie** [reply]
> I think it's best to try to order them in the NFC form (so what we output is properly canonical ordered canonically Unicode :) @Frenzie : do you agree?

If you mean that's doable, yes, that sounds preferable. I haven't looked into it.

**Reply by ichnilatis-gr** [reply]
The problem is that I have no idea of all this (NFC, HEX, DECIMAL....). I just want to update the Greek keyboard and put the combining characters for Greek accents and breathings in the pop-up menu of number 6, so that someone can write all possible combinations of Greek characters with diacritics. Isn't it enough to simply replace the combining characters (codepoints) on the English keyboard with the Greek combining characters?

**Reply by poire-z** [reply]
> Isn't it enough to simply replace the combining characters (codepoints) on the English keyboard with the Greek combining characters?

Yes, indeed, it is enough - if you're fine with the 11 available slots. And just one single diacritic code per slot. Then, the user can himself combine a letter and 1 or more diacritics to build the character he wants.

(If you additionally wanted to have ready made output of the common combinations of ϵ + diacritics when you long press on the <kbd>ϵ</kbd> key, there you would need to build the combinations yourself in the popup for the letter ϵ.)

> The problem is that I have no idea of all this (NFC, HEX, DECIMAL....).

Learning is nice :)

**Reply by ichnilatis-gr** [reply]
Thank you very much!!!
Indeed, I still have a lot to learn... But I still can't make sense of it all... :)

**Comment by ichnilatis-gr** [comment]
While working on upgrading the Greek keyboard, I found that when I search using the combining characters for the diacritics, the search does not recognise the characters that have those diacritics.
e.g. When I search for the word "hellénique" using the "e" and "◌́" keys for the letter "é", the search doesn't recognise the word...

**Reply by poire-z** [reply]
Yes... That's what the "normalizations" I mentionned above are useful - but we currently don't support/do that. May be one day (a bit related: #10366).
So, if you input e + ◌́ , it won't find é in the book.
So, if you input é, it won't find e + ◌́ in the book.
Dunno if it could be enough to normalize the input text (so your e + ◌́ becomes é) before doing the search, assuming/expecting the books to be normalized.

**Reply by ichnilatis-gr** [reply]
That means that there should be deferent keys for each combination of Greek letters and diacritics...
I have almost prepared a new version of the Greek keyboard, based on the English one. I've just changed the English letters to Greek and the diacritics in the pop-up menu of No 6.
Only two keys are available (since the Greek alphabet has 24 letters); those of "Q" and "W". But I don't think that are enough to add all the combinations in their pop-up menus... On the other hand, I wouldn't want to change the "symbolmode" keys...
Maybe I should keep the old layout for the Greek keyoboard?

**Reply by poire-z** [reply]
> That means that there should be deferent keys for each combination of Greek letters and diacritics...

Currently, may be, but it would be bothering, as you would have to know that fact, and to try typing all combinations if you were looking for some exact word.
This shouldn't be the way to handle it, and we will have to fix that. Dunno when.
But I guess people mostly search by highlighting a word, and want to find that word in the book, and hopefully, that same word is written the same way in all that book. That is, I NEVER had to write the letter `é` when doing a search in a French book.

> I have almost prepared a new version of the Greek keyboard, based on the English one. I've just changed the English letters to Greek and the diacritics in the pop-up menu of No 6.
> Maybe I should keep the old layout for the Greek keyoboard?

I have no idea, that Greek keyboard is up to you and other Greek users.
But we have the same "problem" with the French keyboard, and somehow, we have diacritics on the 6 popup (outputing e + ◌́ - but we also have shortcuts in the popup of `a e o u i` allowing to output directly é or à or ï. So, I guess that you should do the same with the Greek keyboard: the 6 popup is to allow any rare combinations, but the popups on all other simple letter keys are there for shortcuts (for the best: the normalized versions of that letter).

**Reply by ichnilatis-gr** [reply]
> So, I guess that you should do the same with the Greek keyboard: the 6 popup is to allow any rare combinations, but the popups on all other simple letter keys are there for shortcuts (for the best: the normalized versions of that letter).

I've already done it, but they are not enough... :-)

**Reply by poire-z** [reply]
Just to see how good enough this would be, normalizing the input (so that if you input e + ◌́, we will search for é, hoping the book is normalized), could you try adding this one line (without the `+` at start):
```diff
--- a/frontend/apps/reader/modules/readersearch.lua
+++ b/frontend/apps/reader/modules/readersearch.lua
@@ -98,6 +98,7 @@ function ReaderSearch:searchCallback(reverse)
         UIManager:show(InfoMessage:new{ text = error_message })
     else
         UIManager:close(self.input_dialog)
+        search_text = require("ffi/utf8proc").normalize_NFC(search_text)
         self:onShowSearchDialog(search_text, reverse, self.use_regex, self.case_insensitive)
     end
 end
```

**Reply by ichnilatis-gr** [reply]
There are combining characters that get normalized and other that don't.
E.g. the " Combining Cyrillic Dasia Pneumata" and "Combining Cyrillic Psili Pneumata" do not get normalized and can't be combined with other combining characters, such as Grave accend or Acute accent.

**Reply by poire-z** [reply]
Tried my one line addition ?
Was your comment about this one line addition causing issues ? Or just some generic comment ?

If they don't usually combine, there is no "one char" unicode codepoint (like `é`), so the normalization won't be able to combine them into a "one char" variant, and they will stay 2 chars - and so will they stay if we normalize the same way text from the book, and we would find a match.
(Then if they combine or not is up to the font and the font rendering engine).

**Reply by ichnilatis-gr** [reply]
My comment is indeed about this one line addition. As for these combining characters, there are unicode codepoints and I have put them in the vowel pop-up menus.

**Reply by poire-z** [reply]
So, please say what you typed/inputed, (and what it became after normalisation if you know, otherwise I'll find out), and what you expected to find in the book that it did not find (screenshot of View HTML after selecting some bit of text containing the expected match).

**Reply by ichnilatis-gr** [reply]
What puzzles me about these two characters ("Combining Cyrillic Dasia Pneumata" and "Combining Cyrillic Psili Pneumata") is that they are named "Cyrillic", so they are probably not the Greek characters of Dasia and Psili breathings, even though they are also marked with that name. The problem with normalization is only encountered with these characters. Unfortunately, I didn't manage to find the Greek ones.

Anyway, one could use the combination of the desired letter with one of these breathings from the pop-up menu in the relevant letter and then use one of the combining characters to add e.g. a tone mark.

**Reply by ichnilatis-gr** [reply]
@poire-z: Is there any thought of merging this line into the code? It looks like it works.

**Reply by poire-z** [reply]
That bit of code was mostly to get an idea if normalization would solve your use case (inputing diacritics in any order on your Greek keyboard).
Using it as is could cause issues, ie. when on a book with NOT normalized content (dunno how common this could be), selecting/highlighting text from the book to search for that text in the book, and then editing that text to clean it up, normalizing it would cause no other text to be found :/
When I looked at it, there was no obvious location where I could do this normalization without causing such side effects, so not sure it would be for the best.

I wrote a bit more some thoughts about how we could get full text search to work properly at https://github.com/koreader/koreader/issues/10366#issuecomment-1586197165 . Not sure if it will come or will come soon, but I'm regularly (when counting sheeps to fall asleep doesn't work :) thinking about that.

**Reply by ichnilatis-gr** [reply]
I appreciate your efforts and thank you very much. I will continue to add this line manually until a solution is found.

**Reply by poire-z** [reply]
Thinking again, my reflexions in the linked issue are about crengine documents. Your problem should also appear when using your keyboard on PDF (where I don't think we'll ever get "good" fulltext seach).
So, I guess I should find a generic frontend solution to normalize the search pattern when inputed from the keyboard (and leave it alone when not).

**Reply by ichnilatis-gr** [reply]
I'll wait in the hope that you won't need to lose sleep too many times, or at least that it will help you sleep more easily :-)
After all, I can't make a PR for an updated Greek keyboard without this problem having been solved...
Thanks!

**Reply by poire-z** [reply]
OK, normalization of the input done with https://github.com/koreader/koreader/pull/10637
